### ISSUE TITLE
[Nextcloud] feature requests

### ISSUE THREAD

**Issue by the-djmaze** [the-djmaze/snappymail#569]
Files
- [x] Save full email
- [x] Save attachments
- [ ] Directory tree selector (work in progress)
- [ ] Attach files to new emails

Contacts
- [x] Search
- [ ] Save

Calendar
- [ ] Add mail message .ics attachment
- [ ] Reply approved/denied calendar item

Related issues:
https://github.com/the-djmaze/snappymail/issues/96
https://github.com/the-djmaze/snappymail/issues/568
https://github.com/the-djmaze/snappymail/discussions/563
https://github.com/the-djmaze/snappymail/discussions/556#discussioncomment-3913855

**Comment by relikd** [comment]
Thank you again for your time and effort. Maybe add the "Help needed" notice banner to your Readme? Hopefully more people will find it that way. I would like to help but I am busy all the time :/ Still, maybe I find some time to help on an issue/feature.

**Comment by Mer0me** [comment]
I'm about to make (and the post here) a full comparison of 3 mail clients usable with nextcloud : Nextcloud official Mail app, Snappy mail and Thunderbird (this one is not really a webmail into nextcloud, but it has some specific features very useful for power users).

For now, each solution lacks at least one major feature, and proposes at least one killer functionality. That makes the choice very difficult.

Anyway, the work accomplished here by the-djmaze is awesome. Thank you so much...

**Comment by Mer0me** [comment]
"Attach files to new emails" is checked, but I can't figure out how to do that. What am I missing ?

**Comment by the-djmaze** [comment]
> I'm about to make (and then post here) a full comparison of 3 mail clients

Great, looking forward to it.

> "Attach files to new emails" is checked, but I can't figure out how to do that. What am I missing ?
You are missing the button right from the paperclip which will be added by plugin version 2.6
![afbeelding](https://user-images.githubusercontent.com/3752035/197198056-767e1f8e-37be-4147-9c80-33c679de3523.png)

**Comment by the-djmaze** [comment]
> But now, it works. Cool.

Good!
The window is still ugly, but it works.

**Comment by Mer0me** [comment]
The address book of Snappy Mail is litterraly replaced by the Nextcloud one ? oO

**Comment by the-djmaze** [comment]
@Mer0me
> .EML opening

SnappyMail can. But only when it is a message attachment.
All done through JavaScript because of OpenPGP PGP/MIME support.
That's how SnappyMail supports PGP/MIME and Nextcloud Mail doesn't

Features | Nextcloud Mail 2.0.3 | SnappyMail 2.19.1 | Thunderbird
-- | -- | -- | --
  |   |   |  
PGP/Inline | Yes (mailvelope) | Yes | Yes
PGP/MIME | No | Yes | Yes
S/MIME | No | No #259 | Yes

About OpenPGP see #89

**Comment by Mer0me** [comment]
Showing these 3 mail solutions to different users, the first reaction is : Netxcloud Mail UI is very nice. That make them forget all the technical limitations of the app :/
My advice : work on the CSS of Snappy !

**Comment by MI-KY** [comment]
First big thanks to @the-djmaze for all the work on the Nextcloud integration, this is really great!

Regarding the comparison table: Nextcloud Mail is able to search for the sender if you use the keywords described here: https://github.com/nextcloud/mail/issues/5611#issuecomment-934477152
But: the search in Nextcloud Mail isn't very good because it is done using the DB of Nextcloud instead of doing an IMAP Search. Therefore (and for many other things) the Nextcloud Mail Client is not a very good alternative to SnappyMail at the moment...

The UI of Nextcloud Mail is nice but in my opinion this has to do with the fact that there are missing functionalities - if Nextcloud Mail would have the same functionalities and options as SnappyMail there would have to be much more buttons ecc. :slightly_smiling_face: . Yes, a different CSS would be nice - but having a better NC integration is even nicer :smile: .

**Comment by Mer0me** [comment]
@MI-KY : thank you for the tip (from:), I saw that an advanced search dialog box is in progress [here](https://github.com/nextcloud/mail/pull/6546)
I'm not sure the UI integration is related to the number of buttons and features. 

I've updated the comparison table and posted it on the Nextcloud github side this time : https://github.com/nextcloud/mail/issues/7482

**Comment by the-djmaze** [comment]
I've released the v2.7 extension for integration.
Go to the SnappyMail Admin panel -> Extensions and update to v2.7

When you click on the cog wheel it should have a new option `Enable "Put ICS in calendar"` for the experimental integration (due to a reported bug).

**Comment by Mer0me** [comment]
> I've added the "Save to calendar" but it crashes Nextcloud Calendar.
> 
> So i've disabled the feature for now as Nextcloud Calendar is not able to process ics files of other applications properly.
> 
> There is also a problem in Sabre for vCards

If it can help :
This app can sync Outlook calendar with Nextcloud Agenda : https://caldavsynchronizer.org/
and this one from Thunderbird Lightning to Nextcloud Agenda : https://addons.thunderbird.net/fr/thunderbird/addon/dav-4-tbsync/
And Nextcloud Agenda can import ICS itself.

What do you mean by "Nextcloud Calendar is not able to process ics files of other applications properly"

**Comment by the-djmaze** [comment]
> What do you mean by "Nextcloud Calendar is not able to process ics files of other applications properly"

The problem is reported at https://github.com/nextcloud/calendar/issues/4684

**Comment by derekakelly** [comment]
I think one feature you may want to consider adding is an unread email widget on the dashboard?

**Comment by hampoelz** [comment]
Nextcloud Theme for Snappymail is on the way!

A pull request will follow soon... The settings and admin panels still needs to be themed, various other small things are still missing and the code needs to be cleaned up.

I've attached some screenshots for a first look.

---

<details>
  <summary>Screenshot with current SnappyMail NC theme</summary>
  
  ![before](https://user-images.githubusercontent.com/28979769/198405555-d8543126-ac83-468c-ae21-a58d15dc9735.png)

</details>

---

**Screenshots with "new" theme:**

![dark-color1](https://user-images.githubusercontent.com/28979769/198404014-7a0cfb2c-3584-47cf-b866-dc697f476429.png)
![light-color1](https://user-images.githubusercontent.com/28979769/198404022-41792cad-ace1-46e2-91b2-74bd8c7ab0ee.png)

<details>
  <summary>(Screenshots with different Nextcloud theme color)</summary>
  
  ![dark-color2](https://user-images.githubusercontent.com/28979769/198404028-9ab7ed31-1e55-4626-bb13-e07330ee2248.png)
  ![light-color2](https://user-images.githubusercontent.com/28979769/198404037-822d8345-3987-4d90-a720-6e0a97090944.png)
  
</details>

**Comment by brendan-pike** [comment]
Is there anything I can do to improve performance of snappy on nextcloud? It's better than nextcloud mail but still takes 12seconds first load snapppy, then 2.7s to load each new email.  I'm wondering if there is a mail inbox pre-fetch function so it displays instantly?

**Comment by elhananjair** [comment]
@hampoelz woow just woow, It's so beautiful. I am eagerly waiting for the merging of your pull request 😊

**Comment by relikd** [comment]
I had to restore my totally broken nextcloud installation during the last 3 days ... so I can't say whether the following error is related to that or whether this is an actual issue with Snappy.

```
An error occurred.
Please refresh the page and try again. Error: Failed loading /extra-apps/snappymail/app/snappymail/v/2.19.5/static/js/min/libs.min.js
```
console says `Uncaught (in promise)` in run on line 21.

(I did move from a snapd to docker installation, so I wouldnt be surprised if something else broke. But so far everything else is working except SnappyMail)

**Comment by relikd** [comment]
> (I did move from a snapd to docker installation, so I wouldnt be surprised if something else broke. But so far everything else is working except SnappyMail)

Nevermind. It was indeed a merge issue. Deleted the `app_path` variable and it was auto-generated again with the proper value. (rename "/extra-apps/" -> "/custom_apps/")

**Comment by brendan-pike** [comment]
Can I suggest that you include version update info in the nextcloud changelog.  It's helpful when updates appear in nextcloud to know what the changes are without having to explore github.  :+1:

**Comment by the-djmaze** [comment]
@brendan-pike how to do that?

**Comment by elhananjair** [comment]
> @brendan-pike how to do that?

https://nextcloudappstore.readthedocs.io/en/latest/developer.html#changelog

**Comment by relikd** [comment]
I'd like to request to change the name of the app in the header area "SnappyMail" -> "E-Mail". (so I dont need to exlpain to non-technical users where to find their emails.

**Comment by rcsilva83** [comment]
@Mer0me
> .ICS to Nextcloud calendar
Is tit already implemented? I tested it but there was no "Accept/Deny" buttons on the invitation e-mail.

**Comment by MI-KY** [comment]
As far as I know this is on the list, but not implemented at the moment - you can see the progress in the first post.
I think accepting invitations can only work when the general problem with ics import has been solved...

**Comment by MI-KY** [comment]
> @MI-KY
> 
> > I think accepting invitations can only work when the general problem with ics import has been solved...
> 
> You mean this issue, right? [nextcloud/calendar#4684](https://github.com/nextcloud/calendar/issues/4684)

Yes. At least to me this seems connected...

**Comment by relikd** [comment]
> Is it possible (maybe already and I'm missing it) to have a Nextcloud file linked rather than attached? Thereby automatically creating a share in Files as well?

Uff, if there was, I would argue strongly against it. Security wise, the sharing dialog is an active part to prevent you from accidentially leaking private information. I would hate it if a mail application would create a share (potentially without password) and leak that information directly via email to another party. Even worse if the email user did not intend to share the Nextcloud link instead of an attachement (document at a specific time vs. automatically getting update.

Sure we can argue that it will make it more convenient, but the more SnappyMail adds on functionality, the more support is needed for issues that arise from it. If you want that feature, you (or someone else) can create a SnappyMail plugin for that instead, maybe?

**Comment by BiglifeMatt** [comment]
@relikd I get what you are saying.  I'm thinking about the overhead of attachments in an email as opposed to the singular file existing.  I think it would be better for the change in culture to handle file shares only from Files and not within email whatsoever.  No offense to Snappymail or any other service but I wonder if the day will arrive when email itself is a thing of the past.  Until then Snappymail is EXCELLENT for my Nextcloud needs!

**Comment by Mer0me** [comment]
@relikd : a good mail application should propose both : link or attached file. When you use a Nextcloud instance with hundred or thousands of corporate users, it's important to prevent users to attach files they want to share, making grow the mailboxes, multiplying versions of a same document, confusing everyone.
Nextcloud official Mail app allows to share a link or to attach the files from Files App.

**Comment by the-djmaze** [comment]
P.S. Rereading the specs, `content-location` should only be used as "alternate location" and the file should still be attached.
Outlook only supports as `filename.URL` attachment (hello TNEF).

So, only links in plain/HTML will always work for public/private file shares.

**Comment by jdaviescoates** [comment]
I find the font size a bit small, be great to have a setting where I can increase that

**Comment by cm-schl** [comment]
> I find the font size a bit small, be great to have a setting where I can increase that

This could give you some help: https://github.com/the-djmaze/snappymail/issues/562
Maybe when the new theme https://github.com/the-djmaze/snappymail/pull/629 is ready your problem could be solved 🙂

**Comment by Mer0me** [comment]
> @BiglifeMatt, @relikd and @Mer0me i've added file sharing.
> 
> Password protect detection is not yet implemented, but Nextcloud can

No time to investigate now, but a simple update of the app on Nextcloud gave me this message :

"An error occurred.
Please refresh the page and try again. Failed loading /apps/snappymail/app/snappymail/v/2.21.1/static/js/min/libs.min.js"

SnappyMail is broken now.

**Comment by the-djmaze** [comment]
@Mer0me which Nextcloud?
There are CSP issues but should be solved.

Best to check your browser console why loading failed.

**Comment by Mer0me** [comment]
> @Mer0me which Nextcloud? There are CSP issues but should be solved.
> 
> Best to check your browser console why loading failed.

Uh ? It works today, maybe a cache problem. Nevermind, sorry.

**Comment by Mer0me** [comment]
> @BiglifeMatt, @relikd and @Mer0me i've added file sharing.
> 
> Password protect detection is not yet implemented, but Nextcloud can

Where is file sharing option ?

**Comment by the-djmaze** [comment]
Contacts avatars now working with the `Avatars` extension.

**Comment by jdaviescoates** [comment]
Latest update to SnappyMail Nextcloud app means that read messages are not marked as read. Standalone version works fine and after reading there they are correctly show as read in Nextcloud too, but if I just read a message within Nextcloud it doesn't show as read, it stays showing as unread.

**Comment by jdaviescoates** [comment]
Also, in SnappyMail Nextcloud app (but not standalone SnappyMail) when I click new message the cursor is in the Body field instead of the To field which is a little annoying :) please could we have it automatically go into the To field? Thanks!

**Comment by jdaviescoates** [comment]
> Latest update to SnappyMail Nextcloud app means that read messages are not marked as read. Standalone version works fine and after reading there they are correctly show as read in Nextcloud too, but if I just read a message within Nextcloud it doesn't show as read, it stays showing as unread.

This is still an issue for me.

**Comment by MI-KY** [comment]
> > Latest update to SnappyMail Nextcloud app means that read messages are not marked as read. Standalone version works fine and after reading there they are correctly show as read in Nextcloud too, but if I just read a message within Nextcloud it doesn't show as read, it stays showing as unread.
> 
> This is still an issue for me.

I don't have this problem on my Nextcloud (v24)... Are you using v25 of Nextcloud? Do you use the iframe or embedded mode of SnappyMail?

**Comment by jdaviescoates** [comment]
> Are you using v25 of Nextcloud?

Yes. I'm on 25.0.2

> Do you use the iframe or embedded mode of SnappyMail?

Embedded

**Comment by cm-schl** [comment]
> > Are you using v25 of Nextcloud?
> 
> Yes. I'm on 25.0.2
> 
> > Do you use the iframe or embedded mode of SnappyMail?
> 
> Embedded

Can't confirm this on the same version of Nextcloud and same embedded mode.
Maybe have a look what you configured in the Admin Panel -> Config -> message_read_delay.
If the problem persists it would be better to discuss this in a separate issue as this is the issue for feature requests...

**Comment by HeyVern** [comment]
Would it be possible to add the handling of mailto:// links for the NextCloud hosted version of snappymail?  The non-Nextcloud version prompts for handling mailto links, but I can't seem to get the NextCloud version to do so.  Thanks!

**Comment by cm-schl** [comment]
> Would it be possible to add the handling of mailto:// links for the NextCloud hosted version of snappymail? The non-Nextcloud version prompts for handling mailto links, but I can't seem to get the NextCloud version to do so. Thanks!

This should already be possible - at least in older versions the question was shown in Firefox. What browser are you using?

**Comment by HeyVern** [comment]
Chrome.  In the stand-alone version of snappymail, it works as expected; however, in the Nextclound integrated version (my primary use case now), it never prompts for being the handler for mailto:.

**Comment by BiglifeMatt** [comment]
Deck is replacing Trello for our company.  It would be nice to have a function to create a Deck Card from an email.

**Comment by the-djmaze** [comment]
@BiglifeMatt 

What i don't understand is why Nextcloud has Calendar and Deck as separate apps.
iCalendar (VCALENDAR) has several objects, including:
- VEVENT (calendar)
- VTODO (tasks)

In the past i wrote a project management tool using VTODO and that worked nicely with other apps.
It allows you to make a whole "waterfall" tree of tasks and calendar events.
It could even add VALARM and the tool would sound the alarm if the VTODO was not 100%.

Since i don't have any clue how Nextcloud Deck works, someone should do the coding.
As a note: even the integration with Nextcloud Calendar is still buggy and crashes because Nextcloud Calendar does not properly support VEVENT

**Comment by dklueh79** [comment]
Integration for Nextcloud Office would be great- in order to open docs etc. directly without saving and going to files...

**Comment by dklueh79** [comment]
> Deck is replacing Trello for our company. It would be nice to have a function to create a Deck Card from an email.

Yes, that would be a great feature

**Comment by relikd** [comment]
@eugenov Do you know that you can add additional accounts to a single "email login"? That way you can log in with a single email address and select the others afterwards in a drop down menu. You can combine that with storing your login credentials in the profile settings. This way you are always logged in if you log in with your nextcloud account.

**Comment by relikd** [comment]
no, the secondary account wont mount the others. Logically, this is a different account and can have different attached secondaries. I know, it just feels wrong :D. ... But in my case, I just created a new email alias and use that for log-in. This new alias is not used for email communication, just for log-in.

**Comment by the-djmaze** [comment]
@eugenov 
You are John and you have a colleague Jane.
1. Jane only has access to her mail.
2. John has access to mail of Jane and himself

When John logs in as Jane he can't see mail of John due to rule 1.
When John logs in as John he can see mail of Jane due to rule 2.

What is the unlogical about this?

**Comment by relikd** [comment]
@the-djmaze Here is another example: John and Jane work at a company and have both access to it-support\<@>company.com. John logs in and adds his private email account john\<@>company.com. Now, Jane has also access to John's mails.

I know, a workaround is to add the private email first and attach the other. But that misses the point that individual actions have impact on other users. An ideal solution for my use case would be to have an administrator view to define shared email accounts and add them to a Nextcloud user (with automatic login). My workaround with aliases is cumbersome but it works, so I wont complain about the current behavior ;-)

**Comment by the-djmaze** [comment]
> Unlogical is to not use local NC account as primary and choose one of secondaries as primary

Nextcloud -> Admin -> Additional Settings -> SnappyMail
There's the option to login automatic with Nextcloud login

**Comment by the-djmaze** [comment]
@eugenov open a discussion for help. This is not the place for your configuration issues.
SnappyMail can and you just need to setup your mail server and nextcloud and snappymail properly.

**Comment by n3ovo** [comment]
Feature Request .. I know that (well I think) signatures are in the "stand alone" version  ... Could there be any plans on adding this to the Nextcloud App version. I know there's  a lot of work going on. But this is something that is missing with out signatures. I keep getting asked for it over and over.

Any thoughts

Thank You ... Very Very Much For All That You All Do......
John G.

**Comment by BiglifeMatt** [comment]
@n3ovo I may be misunderstanding you but I believe that signatures has been a feature for some time now.  You just go into settings within the Snappymail app, under the General tab, you click on the Identity name (username/email address), and a box pops up to enter your signature.   Even excepts formatting and images.

**Comment by n3ovo** [comment]
thank you

**Comment by rcsilva83** [comment]
About the feature of adding an event to calendar, it seams to be blocked by https://github.com/nextcloud/calendar/issues/4684 issue, right? But I was thinking... How Nextcloud Mail app can do this and Snappymail can't?

I may (probably) be very mistaken but wouldn't be the case of copying [this code](https://github.com/nextcloud/mail/blob/b1b1f50d2d94fb1b8649dd9a89504f3a0a13696f/src/components/itinerary/EventReservation.vue#L105-L143) into Snappymail?

**Comment by the-djmaze** [comment]
@rcsilva83 will not work.
SnappyMail imports the ICS as-is.
This means that the ics attached to an email is 1-1 copied.

Sabre and Nextcloud have bugs in parsing valid ics files.

**Comment by akhil1508** [comment]
@the-djmaze Is it possible to have a feature where new contacts are directly saved into nextcloud using the contacts manager just like we do for [contacts suggestions](https://github.com/the-djmaze/snappymail/blob/master/plugins/nextcloud/NextcloudContactsSuggestions.php)? At the same time, disabling snappymail contacts database fully..

This would make the UX of cloud integration much smoother(it can even be optional) like Google Contacts + Gmail for example. 

If you could tell me if this is even possible and give some direction as to how, I would be more than happy to develop it :)

**Comment by aukfood** [comment]
Hello, there is no solution to integrate ics file in snappymail to manage witj nextcloud calendar ?
